File: packages/apple/Sources/Core/OutputParser.swift
import Foundation

// MARK: - OutputParser Protocol

/// Protocol for parsing model output text into structured types.
///
/// ```swift
/// let parser = JSONOutputParser<MyResponse>()
/// let result = try parser.parse(modelOutput)
/// ```
public protocol OutputParser<Output>: Sendable {
    associatedtype Output: Sendable

    /// Instructions to include in the prompt to guide model output format
    var formatInstructions: String { get }

    /// Parse raw model output text into the target type
    func parse(_ text: String) throws -> Output
}

// MARK: - JSON Output Parser

/// Parses JSON-formatted model output into Decodable types
public struct JSONOutputParser<T: Decodable & Sendable>: OutputParser {
    public typealias Output = T

    public var formatInstructions: String {
        "Respond ONLY with valid JSON matching the expected schema. No explanations."
    }

    public init() {}

    public func parse(_ text: String) throws -> T {
        let cleaned = text
            .replacingOccurrences(of: "```json", with: "")
            .replacingOccurrences(of: "```", with: "")
            .trimmingCharacters(in: .whitespacesAndNewlines)

        guard let data = cleaned.data(using: .utf8) else {
            throw LocanaraError.executionFailed("Could not encode response as UTF-8")
        }
        return try JSONDecoder().decode(T.self, from: data)
    }
}

// MARK: - Text Output Parser

/// Passthrough parser for plain text responses
public struct TextOutputParser: OutputParser {
    public typealias Output = String

    public var formatInstructions: String { "" }

    public init() {}

    public func parse(_ text: String) throws -> String {
        text.trimmingCharacters(in: .whitespacesAndNewlines)
    }
}

// MARK: - List Output Parser

/// Parses delimited list responses into arrays
public struct ListOutputParser: OutputParser {
    public typealias Output = [String]

    public let delimiter: String

    public var formatInstructions: String {
        "Return items separated by '\(delimiter)', one per line."
    }

    public init(delimiter: String = "\n") {
        self.delimiter = delimiter
    }

    public func parse(_ text: String) throws -> [String] {
        text.components(separatedBy: delimiter)
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
    }
}

File: packages/android/locanara/src/main/kotlin/com/locanara/core/OutputParser.kt
package com.locanara.core

import kotlinx.serialization.KSerializer
import kotlinx.serialization.json.Json

/**
 * Parses model output text into structured types.
 *
 * ```kotlin
 * val parser = JSONOutputParser(MyResponse.serializer())
 * val result = parser.parse(modelOutput)
 * ```
 */
interface OutputParser<T> {
    /** Instructions to include in the prompt to guide output format */
    val formatInstructions: String

    /** Parse raw model output text into the target type */
    fun parse(text: String): T
}

/**
 * Parses JSON output into serializable types
 */
class JSONOutputParser<T>(
    private val serializer: KSerializer<T>
) : OutputParser<T> {

    private val json = Json { ignoreUnknownKeys = true; isLenient = true }

    override val formatInstructions: String =
        "Respond ONLY with valid JSON matching the expected schema. No explanations."

    override fun parse(text: String): T {
        val cleaned = text
            .replace(Regex("```(?:json)?\\s*"), "")
            .replace("```", "")
            .trim()
        return json.decodeFromString(serializer, cleaned)
    }
}

/**
 * Passthrough parser for plain text
 */
class TextOutputParser : OutputParser<String> {
    override val formatInstructions: String = ""
    override fun parse(text: String): String = text.trim()
}

/**
 * Parses delimited list responses
 */
class ListOutputParser(
    private val delimiter: String = "\n"
) : OutputParser<List<String>> {
    override val formatInstructions: String =
        "Return items separated by '$delimiter', one per line."

    override fun parse(text: String): List<String> =
        text.split(delimiter).map { it.trim() }.filter { it.isNotEmpty() }
}
